src/main/java/org/sonarqube/gradle/SourceCollector.java

```diff
@@ -101,19 +112,27 @@ public SourceCollector(Set<Path> existingSources, Set<Path> directoriesToIgnore,
 
   @Override
   public FileVisitResult preVisitDirectory(Path path, BasicFileAttributes basicFileAttributes) throws IOException {
-    if (
-      isHidden(path) ||
-      isExcludedDirectory(path) ||
-      isCoveredByExistingSources(path)
-    ) {
+    boolean isHiddenDirectoryToSkip = isHidden(path) && !isChildOrGrandChildOfRoot(path);
+
+    if (isHiddenDirectoryToSkip || isExcludedDirectory(path) || isCoveredByExistingSources(path)) {
       return FileVisitResult.SKIP_SUBTREE;
     }
     return FileVisitResult.CONTINUE;
   }
 
-  private static boolean isHidden(Path path) {
-    return StreamSupport.stream(path.spliterator(), true)
-      .anyMatch(token -> token.toString().startsWith("."));
+  private boolean isHidden(Path path) {
+    Path relativePath = root.relativize(path);
+    while (relativePath != null && !relativePath.equals(root)) {
+      if (relativePath.toFile().isHidden()) {
+        return true;
+      }
+      relativePath = relativePath.getParent();
+    }
+    return false;
+  }
+
+  private boolean isChildOrGrandChildOfRoot(Path path) {
+    return path.getParent() != null && (path.getParent().equals(root) || path.getParent().getParent().equals(root));
```

Comment:
`getParent().getParent()` may be null
```java
    Path root = Path.of("a");
    Path path = Path.of("b/c");
    path.getParent().getParent().equals(root); // NullPointerException
 ```
You can flip equals `root.equals(path.getParent().getParent())`